miio/device.py

```diff
@@ -122,205 +117,11 @@ def __init__(
-            "unknown": 0x00000000,
-            "device_id": self._device_id,
-            "ts": send_ts,
-        }
-
-        msg = {"data": {"value": cmd}, "header": {"value": header}, "checksum": 0}
-        m = Message.build(msg, token=self.token)
-        _LOGGER.debug("%s:%s >>: %s", self.ip, self.port, cmd)
-        if self.debug > 1:
-            _LOGGER.debug(
-                "send (timeout %s): %s",
-                self._timeout,
-                Message.parse(m, token=self.token),
-            )
-
-        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        s.settimeout(self._timeout)
-
-        try:
-            s.sendto(m, (self.ip, self.port))
-        except OSError as ex:
-            _LOGGER.error("failed to send msg: %s", ex)
-            raise DeviceException from ex
-
-        try:
-            data, addr = s.recvfrom(1024)
-            m = Message.parse(data, token=self.token)
-            self._device_ts = m.header.value.ts
-            if self.debug > 1:
-                _LOGGER.debug("recv from %s: %s", addr[0], m)
-
-            self.__id = m.data.value["id"]
-            _LOGGER.debug(
-                "%s:%s (ts: %s, id: %s) << %s",
-                self.ip,
-                self.port,
-                m.header.value.ts,
-                m.data.value["id"],
-                m.data.value,
-            )
-            if "error" in m.data.value:
-                error = m.data.value["error"]
-                if "code" in error and error["code"] == -30001:
-                    raise RecoverableError(error)
-                raise DeviceError(error)
-
-            try:
-                return m.data.value["result"]
-            except KeyError:
-                return m.data.value
-        except construct.core.ChecksumError as ex:
-            raise DeviceException(
-                "Got checksum error which indicates use "
-                "of an invalid token. "
-                "Please check your token!"
-            ) from ex
-        except OSError as ex:
-            if retry_count > 0:
-                _LOGGER.debug(
-                    "Retrying with incremented id, retries left: %s", retry_count
-                )
-                self.__id += 100
-                self._discovered = False
-                return self.send(command, parameters, retry_count - 1)
-
-            _LOGGER.error("Got error when receiving: %s", ex)
-            raise DeviceException("No response from the device") from ex
-
-        except RecoverableError as ex:
-            if retry_count > 0:
-                _LOGGER.debug(
-                    "Retrying to send failed command, retries left: %s", retry_count
-                )
-                return self.send(command, parameters, retry_count - 1)
-
-            _LOGGER.error("Got error when receiving: %s", ex)
-            raise DeviceException("Unable to recover failed command") from ex
+        return self.command_sender.send(command, parameters, retry_count)
```

Comment:
No need to deprecate this, simply keep the docstring and make this pass the parameters to the implementation.

Reply:
My thinking was that the consumers of Device shouldn't be sending "raw commands" and if they really want to do that, they can still do it using the `raw_command()` method which passes it to the `self.protocol.send()`

Reply:
I think that'd be fine, but I'd reword the deprecation message to state that `raw_command` should be used instead. We can drop the `send()` then at some point.

Reply:
✅